csrc/index_compute.cpp

```diff
@@ -3291,81 +3291,598 @@ Val* Index::eye(
 }
 
-Val* Index::cpAsyncBulkIndex(
-    TensorView* gmem_tv,
-    TensorView* smem_tv,
-    TensorView* consumer,
-    Val* mbarrier,
-    const std::vector<kir::ForLoop*>& loops) {
-  using namespace tma;
-
-  bool is_load = (gmem_tv != consumer);
-
-  NVF_ERROR(
-      gmem_tv->getMemoryType() == MemoryType::Global,
-      "cpAsyncBulkIndex is only for global memory tensors");
-
-  int64_t dim = (int64_t)gmem_tv->getMaybeAllocationDomain().size();
-  NVF_ERROR(dim > 0);
-  int64_t itemsize = dataTypeSize(gmem_tv->dtype());
+// Note [Schedule and indexing of TMA]
+//
+// TMA is a hardware feature that allows the GPU to load a tile from a tensor of
+// up to 5D. Note that the dimensionality of the tensor in the eyes of TMA is
+// not necessarily the same as the dimensionality of the tensor in the eyes of
+// the user. For example, it is possible that there is a 1D tensor of shape
+// 1000000 and the user wants to view it as (1000, 1000) and load a tile of (8,
+// 8) from it. Another example is, the user has a tensor of shape (1000, 1000),
+// and as long as the tensor is contiguous, the user can view it as a flattened
+// shape 1000000 tensor, and load a tile of 8 from it. The dimensionality of the
+// tensor in the eyes of TMA is determined by the schedule of the tensor.
+// However, the dimensionality of the tensor in the eyes of TMA can not be
+// smaller than the physical dimensionality of the tensor. For example, if the
+// user has a tensor of shape (128, 128, 128, 128, 128, 128), and the tensor is
+// fully discontiguous, then it is not possible to view it as a <5D tensor, thus
+// TMA can not be used to load/store a tile from it.
+//
+// The gmem tensor of TMA store, or the gmem tensor of TMA load when replayed as
+// consumer, must be scheduled like this:
+//
+//                         ---------------------
+//                         | Allocation Domain |
+//                         ---------------------
+//                           | | | | | | | | |
+//                         IterDomain Expressions
+//                             | | | | | | |
+//                         ---------------------
+//                         |  I1           I2  |
+//                         ---------------------
+//                            |            |
+//                          split        split
+//                          /   \        /   |
+//                         I3   I4      I5   I6
+//                         |    |       |    |
+//                        ...   |      ... split
+//                              |          /   |
+//                              |         I7   I8
+//                              |         |    |
+//                              |         |   ...
+//                              |         |
+//                         IterDomain Expressions
+//                          |    |    |    |    |
+//                         Bulk Bulk Bulk Bulk Bulk
```

Comment:
Just a quick question. This part appears to indicate `I4` and `I7` are used together to generate the five "Bulk" domains. Is that correct? Or, are `I4` and `I7` transformed independently?

Reply:
They can be independent or together, it makes no difference.